File: Algorithm/Greedy/435_non_overlapping_intervals/q435.py
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort()        
        prev_end = intervals[0][1]
        total_operation = 0
        for i in range(1, len(intervals)):
            if prev_end > intervals[i][0]:
                total_operation += 1
                prev_end = intervals[i][1] if intervals[i][1] < prev_end else prev_end
            else:
                prev_end = intervals[i][1]
        return total_operation

File: Algorithm/Greedy/435_non_overlapping_intervals/q435.cpp
class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        if (intervals.empty()) {
            return 0;
        }
        // sort intervals by end time
        sort(intervals.begin(), intervals.end(), [](const vector<int> &a, const vector<int> &b) {
            return a[1] < b[1];
        });
        // greedy: always keep smallest end time intervals
        int total_operation = 0, prev_end = intervals[0][1], size = intervals.size();
        for (int i = 1; i < size; ++i) {
            if (intervals[i][0] < prev_end) {
                ++total_operation;
            } else {
                prev_end = intervals[i][1];
            }
        }
        return total_operation;
    }
};
